include/gsl/gsl

```diff
@@ -74,69 +74,65 @@ class not_null
     static_assert(std::is_assignable<T&, std::nullptr_t>::value, "T cannot be assigned nullptr.");
 
 public:
-    not_null(T t) : ptr_(t) { ensure_invariant(); }
-    not_null& operator=(const T& t)
-    {
-        ptr_ = t;
-        ensure_invariant();
-        return *this;
-    }
+    constexpr not_null(T t) : ptr_(t) { ensure_invariant(); }
 
     not_null(const not_null& other) = default;
     not_null& operator=(const not_null& other) = default;
 
     template <typename U, typename Dummy = std::enable_if_t<std::is_convertible<U, T>::value>>
-    not_null(const not_null<U>& other)
+    constexpr not_null(const not_null<U>& other) : ptr_(other.get())
     {
-        *this = other;
+        ensure_invariant(); // Just in case a non-null U converts to null T.
     }
 
-    template <typename U, typename Dummy = std::enable_if_t<std::is_convertible<U, T>::value>>
-    not_null& operator=(const not_null<U>& other)
+    constexpr T get() const
     {
-        ptr_ = other.get();
-        return *this;
+        // help the optimizer
+        GSL_ASSUME(ptr_ != nullptr);
+        return ptr_;
     }
 
+    constexpr operator T() const { return get(); }
+    constexpr T operator->() const { return get(); }
+
+    // unary +: because there's no particular reason to forbid it,
+    // and it's sometimes useful to convert an lvalue to a prvalue.
```

Comment:
I have no problem with this. I learned something today: that unary `+` is useful to library implementers ;-)